guppylang/definition/common.py

```diff
@@ -0,0 +1,118 @@
+import ast
+import itertools
+from abc import ABC, abstractmethod
+from collections.abc import Iterator
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, ClassVar, TypeAlias
+
+from guppylang.hugr.hugr import Hugr, Node
+
+if TYPE_CHECKING:
+    from guppylang.checker.core import Globals
+    from guppylang.compiler.core import CompiledGlobals
+    from guppylang.module import GuppyModule
+
+
+RawDef: TypeAlias = "ParsableDef | ParsedDef"
+ParsedDef: TypeAlias = "CheckableDef | CheckedDef"
+CheckedDef: TypeAlias = "CompilableDef | CompiledDef"
+
+
+@dataclass(frozen=True)
+class DefId:
+    """Unique identifier for definitions across modules.
+
+    This id is persistent across all compilation stages. It can be used to identify a
+    definition at any step in the compilation pipeline.
+    """
+
+    id: int
+    module: "GuppyModule | None" = field(compare=False, hash=False)
+
+    _ids: ClassVar[Iterator[int]] = itertools.count()
+
+    @classmethod
+    def fresh(cls, module: "GuppyModule | None" = None) -> "DefId":
+        return DefId(next(cls._ids), module)
+
+
+@dataclass(frozen=True)
+class Definition(ABC):
+    """Abstract base class for user-defined objects on module-level.
+
+    Each definition is identified by a globally unique id. Furthermore, we store the
+    user-picked name for the defined object and an optional AST node for the definition
+    location.
+    """
+
+    id: DefId
+    name: str
+    defined_at: ast.FunctionDef | None
+
+    @property
+    @abstractmethod
+    def description(self) -> str:
+        """ """
+
+
+class ParsableDef(Definition, ABC):
```

Comment:
I'm not sure what the point is of inheriting `Definition, ABC` since `Definition` already inherits `ABC`?

Reply:
Fixed